packaging.md

```diff
@@ -0,0 +1,156 @@
+# Packaging Applications for Deployment
+
+Once an application is built for production, it still needs to be packaged before it can be deployed to servers. There are several strategies for packaging Swift applications for deployment.
+
+## Docker
+
+One of the most popular ways to package applications these days is using container technologies such as [Docker](https://www.docker.com).
+
+Using Docker's tooling, we can build and package the application as a Docker image, publish it to a Docker repository, and later launch it directly on a server or on a platform that supports Docker deployments such as [Kubernetes](https://kubernetes.io). Many public cloud providers including AWS, GCP, Azure, IBM and others encourage this kind of deployment.
+
+Here is an example `Dockerfile` that builds and packages the application on top of CentoOS:
+
+```Dockerfile
+#------- build -------
+FROM swiftlang/swift:nightly-centos8 as builder
+
+# set up the workspace
+RUN mkdir /workspace
+WORKDIR /workspace
+
+# copy the source to the docker image
+COPY . /workspace
+
```

Comment:
@tomerd 99.9% sure that since 5.2 we don’t need the `-Xswiftc -g` anymore

Reply:
@aciidb0mb3r?

Reply:
@tomerd yes, I'm correct. See this:

5.1 without `-g` leads to no source info:

```
$ jw-docker-swift-5.1 bash -c "rm -rf .build && swift build -c release  && lldb -o 'break set -n fcntl' -o run -o bt .build/release/tpkg"
[2/2] Linking tpkg
(lldb) target create ".build/release/tpkg"
Current executable set to '.build/release/tpkg' (x86_64).
(lldb) break set -n fcntl
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) run
3 locations added to breakpoint 1
1 location added to breakpoint 1
1 location added to breakpoint 1
Process 56 stopped
* thread #1, name = 'tpkg', stop reason = breakpoint 1.1
    frame #0: 0x00007ffff7fedd00 libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32
libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32:
->  0x7ffff7fedd00 <+0>: pushq  %rbp
    0x7ffff7fedd01 <+1>: movq   %rsp, %rbp
    0x7ffff7fedd04 <+4>: xorl   %eax, %eax
    0x7ffff7fedd06 <+6>: popq   %rbp
Target 0: (tpkg) stopped.

Process 56 launched: '/tmp/tpkg/.build/release/tpkg' (x86_64)
(lldb) bt
* thread #1, name = 'tpkg', stop reason = breakpoint 1.1
  * frame #0: 0x00007ffff7fedd00 libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32
    frame #1: 0x0000555555554a19 tpkg`main + 9
    frame #2: 0x00007ffff655cb97 libc.so.6`__libc_start_main + 231
    frame #3: 0x000055555555479a tpkg`_start + 42
```

but with `-g`

```
$ jw-docker-swift-5.1 bash -c "rm -rf .build && swift build -c release -Xswiftc -g && lldb -o 'break set -n fcntl' -o run -o bt .build/release/tpkg"
[2/2] Linking tpkg
(lldb) target create ".build/release/tpkg"
Current executable set to '.build/release/tpkg' (x86_64).
(lldb) break set -n fcntl
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) run
3 locations added to breakpoint 1
1 location added to breakpoint 1
1 location added to breakpoint 1
Process 56 stopped
* thread #1, name = 'tpkg', stop reason = breakpoint 1.1
    frame #0: 0x00007ffff7fedd00 libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32
libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32:
->  0x7ffff7fedd00 <+0>: pushq  %rbp
    0x7ffff7fedd01 <+1>: movq   %rsp, %rbp
    0x7ffff7fedd04 <+4>: xorl   %eax, %eax
    0x7ffff7fedd06 <+6>: popq   %rbp
Target 0: (tpkg) stopped.

Process 56 launched: '/tmp/tpkg/.build/release/tpkg' (x86_64)
(lldb) bt
* thread #1, name = 'tpkg', stop reason = breakpoint 1.1
  * frame #0: 0x00007ffff7fedd00 libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32
    frame #1: 0x0000555555554a19 tpkg`main at main.swift:8:1 [opt]
    frame #2: 0x00007ffff655cb97 libc.so.6`__libc_start_main + 231
    frame #3: 0x000055555555479a tpkg`_start + 42
```

In 5.2 however, even without `-g`:

```
$ jw-docker-swift-5.2 bash -c "rm -rf .build && swift build -c release  && lldb -o 'break set -n fcntl' -o run -o bt .build/release/tpkg"
[2/2] Linking tpkg
(lldb) target create ".build/release/tpkg"
Current executable set to '/tmp/tpkg/.build/release/tpkg' (x86_64).
(lldb) break set -n fcntl
Breakpoint 1: no locations (pending).
WARNING:  Unable to resolve breakpoint to any actual locations.
(lldb) run
3 locations added to breakpoint 1
1 location added to breakpoint 1
1 location added to breakpoint 1
Process 56 stopped
* thread #1, name = 'tpkg', stop reason = breakpoint 1.1
    frame #0: 0x00007ffff7fe9a00 libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32
libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32:
->  0x7ffff7fe9a00 <+0>: pushq  %rbp
    0x7ffff7fe9a01 <+1>: movq   %rsp, %rbp
    0x7ffff7fe9a04 <+4>: xorl   %eax, %eax
    0x7ffff7fe9a06 <+6>: popq   %rbp
Target 0: (tpkg) stopped.

Process 56 launched: '/tmp/tpkg/.build/release/tpkg' (x86_64)
(lldb) bt
* thread #1, name = 'tpkg', stop reason = breakpoint 1.1
  * frame #0: 0x00007ffff7fe9a00 libswiftGlibc.so`Glibc.fcntl(Swift.Int32, Swift.Int32, Swift.Int32) -> Swift.Int32
    frame #1: 0x0000555555554a89 tpkg`main at main.swift:8:1 [opt]
    frame #2: 0x00007ffff6affb97 libc.so.6`__libc_start_main + 231
    frame #3: 0x00005555555547ba tpkg`_start + 42
(lldb) 
```

Nice! Source info and (what you can't see here): even colours in the terminal!

Reply:
@weissi cool, I will remove from there and we should update our readme.md as well with this info.

Reply:
@tomerd done, https://github.com/swift-server/guides/pull/13